Generate GIF without asking for location

sunshinejr opened the pull request:
Fixes #72. This needs us to rebuild our current architecture to use controllers as well so it will be another big PR.

sindresorhus commented:
Keep in mind that all the stages needs to be drop targets. The user should be able to drop a video at the initial view, when showing the settings, and when the conversion is completed.

sindresorhus commented:
Some quick things I noticed. Feel free to ignore them for now as they're nitpick. Just wanted to write it down while I remember.

- The first time you drag and drop, there's no blue highlight around the window contents.
- If I open a file through the "Open" menu item, there's no animated transition to the larger window.
- If a drag in a file, then press Esc, and drag in the file again, the "Convert" button is no longer highlighted (accent color).

sindresorhus commented on Gifski/DropVideoViewController.swift:
```diff
@@ -0,0 +1,113 @@
+
+final class DropVideoViewController: NSViewController {
+	private lazy var videoDropView = with(VideoDropView()) {
+		$0.frame.size = CGSize(width: 360, height: 240)
```
I don't think this should be hard-coded.

sunshinejr replied:
Somehow we need to specify a window size for the controller. I cannot use the size of a current window as this might be e.g. Edit controller that has a different size than most of the controllers. Maybe we would need a constant size for all the controllers except the Edit one?

sindresorhus replied:
Yes, I would prefer defining the size in the view controller view instead of here.

sindresorhus commented on Gifski/DropVideoViewController.swift:
```diff
@@ -0,0 +1,113 @@
+		$0.dropText = "Drop a Video to Convert to GIF"
+		$0.onComplete = { [weak self] url in
+			NSApp.activate(ignoringOtherApps: true)
+			self?.convert(url.first!)
```
`url.first` reads weird. I think we should change the drop view to only return a single URL.

sindresorhus commented on Gifski/DropVideoViewController.swift:
```diff
@@ -0,0 +1,113 @@
+				)
+
+				return
+		}
```
Can you move all these checks into a separate method? `validateInput()` or something.

sindresorhus commented on Gifski/DropVideoViewController.swift:
```diff
@@ -0,0 +1,113 @@
+			let dimensions = asset.dimensions,
+			dimensions.width > 10,
+			dimensions.height > 10
+			else {
```
```suggestion
		else {
```

sindresorhus commented on Gifski/MainWindowController.swift:
```diff
@@ -296,25 +46,23 @@ final class MainWindowController: NSWindowController {
+			window?.contentViewController?.push(viewController: dropVideo) {
+				dropVideo.convert(inputUrl)
+			}
 		}
```
Is there really no better way of doing this?

sunshinejr replied:
There must be, we don't need the drop video controller here 😄 Will be fixed.

sindresorhus commented on Gifski/util.swift:
```diff
@@ -2417,3 +2418,45 @@ extension URL {
+		let newOrigin = CGPoint(x: window.frame.midX - viewController.view.frame.width / 2.0, y: window.frame.midY - viewController.view.frame.height / 2.0)
+		let newWindowFrame = CGRect(origin: newOrigin, size: viewController.view.frame.size)
+
+		let animation = CABasicAnimation()
```
Would be cool to use CASpringAnimation to make the transition even more organic.

sindresorhus commented on Gifski/util.swift:
```diff
@@ -2417,3 +2418,45 @@ extension URL {
+}
+
+extension URL {
+	static func generateTempGifUrl(for videoUrl: URL) -> URL {
```
I don't think this should be an extension on URL. It's too specific to this use-case.

sindresorhus commented on Gifski/util.swift:
```diff
@@ -2417,3 +2418,45 @@ extension URL {
+extension URL {
+	static func generateTempGifUrl(for videoUrl: URL) -> URL {
+		let tempDirectory = FileManager.default.temporaryDirectory
+		let tempName = videoUrl.filenameWithoutExtension + ".\(FileType.gif.fileExtension)"
```
```suggestion
		let tempName = "\(videoUrl.filenameWithoutExtension).\(FileType.gif.fileExtension)"
```

sindresorhus commented:
- There's no longer any animation on the thumbnail when showing the conversion completed view.
- What was the reason for switching to a XIB for the conversion completed view?

sunshinejr commented on Gifski/AppDelegate.swift:
```diff
@@ -26,13 +26,13 @@ final class AppDelegate: NSObject, NSApplicationDelegate {
 
 	func application(_ application: NSApplication, open urls: [URL]) {
-		guard !mainWindowController.isRunning else {
+		guard !(mainWindowController.window?.contentViewController is ConversionViewController) else {
```
yes definitely will replace this line, was just a placeholder

sindresorhus commented on Gifski/Constants.swift:
```diff
@@ -8,3 +8,7 @@ extension Defaults.Keys {
+
+extension CGSize {
+	static let defaultWindowSize = CGSize(width: 360, height: 240)
+}
```
This would make more sense as `NSWindow.defaultWindowSize`.

sindresorhus replied:
Actually, `Constants` would be better, since it's not a generic value.

sunshinejr commented:
@sindresorhus so I've made it a `.xib` as it was really hard to read the logic in there - lots of views, lots of constraints.

sindresorhus commented on Gifski/VideoUrlValidator.swift:
```diff
@@ -0,0 +1,99 @@
+	enum Result {
+		case failure
+		case success(AVURLAsset.VideoMetadata)
+	}
```
Why Result for this? I usually only use Result for async operations. `throws` would be more fitting, but since there are no error message, I would just make the return value optional.

sunshinejr replied:
So I want to validate the video but also return `videoMetadata`. The result of the `validate` operation should still be kind of a `Boolean` and so I'm using `videoMetadata` as an associated value to the success/failure.

sindresorhus replied:
Ok. I see your point of wanting to make it a boolean. By making it an optional, I meant that I would just rename it to `getVideoMetadata` or something which also validates, but your solution is cleaner, so let's keep it as is.

sindresorhus commented on Gifski/VideoUrlValidator.swift:
```diff
@@ -0,0 +1,99 @@
+import Crashlytics
+import AVKit
+
+struct VideoUrlValidator {
```
It's not just validating the URL, so this name is a bit misleading. `VideoValidator`? Does it even need to be wrapped in a struct?

sunshinejr replied:
👍 good call on the name - do you have any other idea than using a struct for that? I wanted to be able to extract the logic for the validation to a separate place

sindresorhus replied:
Nah, it's fine.

sindresorhus commented on Gifski/VideoDropViewController.swift:
```diff
@@ -1,7 +1,7 @@
 
 final class VideoDropViewController: NSViewController {
-	private let urlValidator = VideoUrlValidator()
+	private let urlValidator = VideoValidator()
```
`urlValidator` => `videoValidator`

sunshinejr commented:
I've brought back the thumbnail animation. Also, @sindresorhus, there are the issues you noticed:
- The first time you drag and drop, there's no blue highlight around the window contents.
- If a drag in a file, then press Esc, and drag in the file again, the "Convert" button is no longer highlighted (accent color).

Do you have an idea why would that happen? Seems strange.

Also, I will open the PR for a review now, spent many hours on this one and a fresh pair of eyes would be lovely.

sindresorhus commented:
> The first time you drag and drop, there's no blue highlight around the window contents.

I figured this one out. It's caused by: https://github.com/sindresorhus/Gifski/blob/8cd24a5184a8bcef9e1555c367ad64ac8dc00799/Gifski/MainWindowController.swift#L84

> If a drag in a file, then press Esc, and drag in the file again, the "Convert" button is no longer highlighted (accent color).

This one is probably about first responder. When you press Esc you resign first responder, so the button is no longer blue. Maybe we should have the view controller be first responder when shown.

sunshinejr commented:
@sindresorhus when the controller is pushed [I'm making it the first responder by default](https://github.com/sindresorhus/Gifski/pull/94/files#diff-75a4e71e51a82ac8607443b8f202430eR2443). Also both `ESC` and tapping on `Cancel` button go trouch the `cancel` method that's why it's interesting to me what else does `ESC` do that makes the blue button resign its default state.

sunshinejr commented:
@sindresorhus figured it out. I added a `key equivalent` to the Cancel button and it got in that weird state. Cancel is called anyways on `ESC` click so when I removed it it all works correctly again.

sunshinejr commented:
With the vibrancy thing, any hack to make it work without deleting the line?

sindresorhus commented on Gifski/ConversionViewController.swift:
```diff
@@ -0,0 +1,124 @@
+	private func stopConversion(_ completion: (() -> Void)? = nil) {
+		isRunning = false
+		progress?.cancel()
+		progress?.unpublish()
```
To fix the crash when canceling, you need to do this in a next tick. The reason is that after `.cancel()`, the libgifski side needs to cancel too. Otherwise they will continue sending their last update, but `progress` will no longer exist and it will crash.
```suggestion

		DispatchQueue.main.async {
			self.progress?.unpublish()
		}

```

sunshinejr replied:
Hmm, tried that and still got the crash sometimes 🤔

sindresorhus replied:
What kind of crash though? Maybe it's a different one. Can you share?

sunshinejr replied:
<img width="1220" alt="Zrzut ekranu 2019-07-8 o 14 57 15" src="https://user-images.githubusercontent.com/5232779/60811976-c2f96e80-a190-11e9-8dee-945ae08b1b1a.png">

sindresorhus commented:
> With the vibrancy thing, any hack to make it work without deleting the line?

I think you could use `view.insertVibrancyView(material: .underWindowBackground)` on the view controllers instead.

sindresorhus commented:
I noticed now that the progress circle get's less opacity/color when it shows the checkmark (before fading out). I don't think it was like this before.

sindresorhus commented:
For the conversion completed view, the buttons fade in must earlier and faster than the thumbnail. In the current Gifski version, it's better synced.

sindresorhus commented on Gifski/ConversionCompletedViewController.swift:
```diff
@@ -0,0 +1,159 @@
+		draggableFile.layer?.animateScaleMove(fromScale: 3, fromY: view.frame.height + draggableFile.frame.size.height)
+	}
+
+	private func setupUI() {
```
```suggestion
	private func setUpUI() {
```

sunshinejr replied:
hmm, in the app the convention is always `setup` rather than `setUp` - do you prefer that I change all the places to `setUp` instead?

sindresorhus replied:
Because alone it can be used as a noun `setup`, but when you have `UI` afterwards it needs to be a verb `set up`. It could be `uiSetup`.

https://writingexplained.org/setup-vs-set-up-difference

sindresorhus commented on Gifski/ConversionCompletedViewController.swift:
```diff
@@ -0,0 +1,159 @@
+		}
+	}
+
+	private func setupDropView() {
```
```suggestion
	private func setUpDropView() {
```

sindresorhus commented on Gifski/ConversionViewController.swift:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	override func loadView() {
+		let wrapper = NSView(frame: CGRect(origin: .zero, size: Constants.defaultWindowSize))
```
```suggestion
		let wrapper = NSView(frame: Constants.defaultWindowSize.cgRect)
```

sunshinejr commented:
Hmm, I thought that I fixed the vibrancy by resequencing the lines but it seems that the bug is different  than what we thought - the vibrancy view makes the focus on itself and thus we don't see the blue highlight. It is visible afterwards because the next VC overlaps the vibrancy view and thus there is no vibrancy anymore but we can see the blue highlight.

So I think we should somehow stop the vibrancy view from stealing the focus (perhaps `canBecomeKeyView`?) and then on each VC we should add the vibrancy view probably.

sindresorhus commented:
> So I think we should somehow stop the vibrancy view from stealing the focus (perhaps canBecomeKeyView?) and then on each VC we should add the vibrancy view probably.

Sounds correct in theory, yes. But you never know with Cocoa.

sindresorhus commented:
Nice! I had no idea NSWindow has a built-in vibrancy view.

sindresorhus commented:
I noticed another issue. When it's done converting, the checkmark is drawn above the text `100%`:

<img width="472" alt="Screen Shot 2019-07-08 at 23 51 27" src="https://user-images.githubusercontent.com/170270/60827949-869f3000-a1db-11e9-9d5e-3316c36ee17c.png">

sunshinejr commented:
@sindresorhus yeah I also noticed it and fixed it. Will update the PR shortly. Currently trying to figure out why my memory debugger says we have 70+ leaks currently, even when I replace `window.contentView` to be a simple `NSView` instead of using any of the created controllers at all. I fixed all retain cycles I found (didn't push the commit yet) but still got this weird state and I'm trying to figure out what's going on.

sunshinejr commented:
Okay, it seemed to be a weird Xcode state, I can't reproduce the issue anymore and I don't see any leaks/cycles. Got one issue in the debugger though:

> objc[66756]: __weak variable at 0x102705e30 holds 0x102700190 instead of 0x100f2c120. This is probably incorrect use of objc_storeWeak() and objc_loadWeak(). Break on objc_weak_error to debug.

sunshinejr commented:
Did some more investigation and it seems to be connected to the transition util? When I changed:
```swift
NSAnimationContext.runAnimationGroup({ _ in
	window.contentViewController?.view.animator().alphaValue = 0.0
	window.animator().setFrame(newWindowFrame, display: true)
}, completionHandler: {
	window.contentViewController = viewController
	viewController.view.animator().alphaValue = 1.0
	window.makeFirstResponder(viewController)
	completion?()
})
```

to 
```swift
window.contentViewController?.view.animator().alphaValue = 0.0
window.animator().setFrame(newWindowFrame, display: true)
window.contentViewController = viewController
viewController.view.animator().alphaValue = 1.0
window.makeFirstResponder(viewController)
completion?()
```
then the warning doesn't happen... I did try to debug on given symbol `objc_weak_error` but it doesn't say much to me...

sunshinejr commented:
False alarm, when removed the context it appeared anyways e.g. in the conversion completed view.

sunshinejr commented:
@sindresorhus I give up on the `objc_weak_error` in the console, tested the app for many days now and it doesn't seem to affect anything. I removed `unpublish` and I can't seem to reproduce the crash either.

I also removed some relict Save Panel hacks, fixed a bug when you selected "percent" based gif size and added focus on dimensions drop down in the Edit VC. Please take a look and see what we're missing.

sindresorhus commented:
> I give up on the objc_weak_error in the console, tested the app for many days now and it doesn't seem to affect anything.

Yeah, doesn't matter if it has no effect on anything. I have a feeling it's some kind of Swift bug.

> I removed unpublish and I can't seem to reproduce the crash either.

👌

sindresorhus commented:
There's still

> I noticed now that the progress circle get's less opacity/color when it shows the checkmark (before fading out). I don't think it was like this before.

and

> For the conversion completed view, the buttons fade in must earlier and faster than the thumbnail. In the current Gifski version, it's better synced.

sindresorhus commented:
Quick Look is no longer working. Nothing happens when I press Space and if I select "Quick Look" in the menu, I get:

<img width="912" alt="Screen Shot 2019-07-18 at 17 10 47" src="https://user-images.githubusercontent.com/170270/61449388-0917b480-a97f-11e9-937e-527e118a8dee.png">

sunshinejr commented:
@sindresorhus oh interesting, do you always have this problem with QuickLook? I cannot reproduce it on my videos 🤔

sunshinejr commented:
Hmm, actually, I can still reproduce the crash - after a successful GIF generation, I try another one and Cancel at 4% and it always crashes. Gonna see if I can figure out why is this happening.

sunshinejr commented:
Fixed the cancel crash (I think)! So we were double cancelling and depending if the second cancel was fast enough or not, the data was retained or not, and then crashed or not 😄 

We were double cancelling because cancelling of the progress cancelled the generator -> thus Gifski reported cancelled state and we then cancelled again... I could've skipped the cancelling on Gifski response.cancelled, but I just added a guard to cancel only if it's not cancelled yet, because I'm not sure if there are other cases when we could get cancel event from Gifski.

sindresorhus commented:
Nice catch!

sindresorhus commented:
I noticed one other small problem. One the conversion completed view, if I Command+H to hide the window and the show the window again, the thumbnail animation plays each time. It should only animate in when the view is presented, not when the window is made visible.

sunshinejr commented:
@sindresorhus 
- Conversion completed view should have better animation sync
- the bug with second animation on minimize should be fixed
- the bug with progress circle opacity should be fixed as well

I can't reproduce the issue with QuickLook so no idea how to approach this one...

sindresorhus commented:
Looks great now. Super nice work, @sunshinejr! Gifski is so much more usable now. 🙌
